gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractProguardTask.kt

```diff
@@ -0,0 +1,150 @@
+
+    private val rootConfigurationFile = workingDir.map { it.file("root-config.pro") }
+
+    private val jarsConfigurationFile = workingDir.map { it.file("jars-config.pro") }
+
+    @TaskAction
+    fun execute() {
+        val javaHome = File(javaHome.get())
+        val proguardFiles = project.configurations.detachedConfiguration(
+            project.dependencies.create("com.guardsquare:proguard-gradle:${proguardVersion.get()}")
+        ).files
+
+        cleanDirs(destinationDir, workingDir)
+        val destinationDir = destinationDir.ioFile.absoluteFile
+
+        // todo: can be cached for a jdk
+        val jmods = javaHome.resolve("jmods").walk().filter {
+            it.isFile && it.path.endsWith("jmod", ignoreCase = true)
+        }.toList()
+
+        val inputToOutputJars = LinkedHashMap<File, File>()
+        // avoid mangling mainJar
+        inputToOutputJars[mainJar.ioFile] = mainJarInDestinationDir.ioFile
+        for (inputFile in inputFiles) {
+            if (inputFile.name.endsWith(".jar", ignoreCase = true)) {
+                inputToOutputJars.putIfAbsent(inputFile, destinationDir.resolve(inputFile.mangledName()))
+            } else {
+                inputFile.copyTo(destinationDir.resolve(inputFile.name))
+            }
+        }
+
+        jarsConfigurationFile.ioFile.bufferedWriter().use { writer ->
+            for ((input, output) in inputToOutputJars.entries) {
+                writer.writeLn("-injars '${input.normalizedPath()}'")
+                writer.writeLn("-outjars '${output.normalizedPath()}'")
+            }
+
+            for (jmod in jmods) {
+                writer.writeLn("-libraryjars '${jmod.normalizedPath()}'(!**.jar;!module-info.class)")
+            }
+        }
+
+        rootConfigurationFile.ioFile.bufferedWriter().use { writer ->
+            writer.writeLn("""
+                -keep public class ${mainClass.get()} {
+                    public static void main(java.lang.String[]);
+                }
+            """.trimIndent())
+
+            val includeFiles = sequenceOf(
+                jarsConfigurationFile.ioFile,
+                defaultComposeRulesFile.ioFile
+            ) + configurationFiles.files.asSequence()
+            for (configFile in includeFiles.filterNotNull()) {
+                writer.writeLn("-include '${configFile.normalizedPath()}'")
+            }
+        }
+
+        val javaBinary = jvmToolFile(toolName = "java", javaHome = javaHome)
+        val args = arrayListOf<String>().apply {
+            val maxHeapSize = maxHeapSize.orNull
+            if (maxHeapSize != null) {
+                add("-Xmx:$maxHeapSize")
+            }
+            cliArg("-cp", proguardFiles.map { it.normalizedPath() }.joinToString(File.pathSeparator))
+            add("proguard.ProGuard")
+            // todo: consider separate flag
+            cliArg("-verbose", verbose)
+            cliArg("-include", rootConfigurationFile)
```

Comment:
this //todo comment needs for the future?

Reply:
Yes, that's stuff to consider